### DISCUSSION TITLE
`datetime` cannot be converted to type `isodatetime`

### DISCUSSION THREAD

**Discussion by jccava** [NeurodataWithoutBorders/helpdesk#46] [Q&A]
Hello all,  I updated my MatNWB (2.6.0) and am now encountering an error while trying to create an NWB file.  Simple code looks like:

```
dt = datetime('now', 'TimeZone', 'local');
dt.Format = 'yyyy-MM-DD''T''HH:mm:ss.SSSZ';
dt

nwbOutputFile = NwbFile();
nwbOutputFile.identifier = 'test';
nwbOutputFile.session_description = 'test';
nwbOutputFile.session_start_time = dt;
```

and what Matlab tells me is:

```

dt = 

  datetime

   2023-03-79T12:50:43.727-0400

Error using assert
Value of type `datetime` cannot be converted to type `isodatetime`.

Error in types.util.correctType (line 60)
        assert(isa(val, type), ...

Error in types.util.checkDtype (line 96)
correctedValue = types.util.correctType(val, type);

Error in types.core.NWBFile/validate_session_start_time (line 744)
        val = types.util.checkDtype('session_start_time', 'isodatetime', val);

Error in types.core.NWBFile/set.session_start_time (line 356)
        obj.session_start_time = obj.validate_session_start_time(val);

Error in processKilosortForNWB (line 34)
nwbOutputFile.session_start_time = dt;
```
 It also breaks even with a simple

`nwbOutputFile.session_start_time = datetime(2018, 4, 25, 2, 30, 3);`

Any ideas what I'm doing wrong?   Thanks in advance.   --James

**Comment by bendichter** [comment]
Can you try initiating the NWB File the way it is done in the tutorials? e.g. [here](https://neurodatawithoutborders.github.io/matnwb/tutorials/html/intro.html)

**Reply by jccava** [reply]
Thanks for the quick reply.

Executing:

```
nwb = NwbFile( ...
    'session_description', 'mouse in open exploration',...
    'identifier', 'Mouse5_Day3', ...
    'session_start_time', datetime(2018, 4, 25, 2, 30, 3), ...
    'general_experimenter', 'My Name', ... % optional
    'general_session_id', 'session_1234', ... % optional
    'general_institution', 'University of My Institution', ... % optional
    'general_related_publications', 'DOI:10.1016/j.neuron.2016.12.011'); % optional
```

gives me:

```
Error using assert
Value of type `datetime` cannot be converted to type `isodatetime`.

Error in types.util.correctType (line 60)
        assert(isa(val, type), ...

Error in types.util.checkDtype (line 96)
correctedValue = types.util.correctType(val, type);

Error in types.core.NWBFile/validate_session_start_time (line 744)
        val = types.util.checkDtype('session_start_time', 'isodatetime', val);

Error in types.core.NWBFile/set.session_start_time (line 356)
        obj.session_start_time = obj.validate_session_start_time(val);

Error in types.core.NWBFile (line 177)
        obj.session_start_time = p.Results.session_start_time;

Error in NwbFile (line 16)
            obj = obj@types.core.NWBFile(varargin{:});

Error in filedatatest (line 1)
nwb = NwbFile( ...
```

Did I mess up while upgrading or perhaps miss a step?

--James

**Reply by lawrence-mbf** [reply]
Do you happen to know what version of the schema you have currently generated?

**Reply by jccava** [reply]
I have a 2.6.0 folder with non-empty subfolders in my  nwb-schema folder (along with older ones).

**Reply by lawrence-mbf** [reply]
In your script, can you run `nwbOutputFile.version` and provide the output of that? I am unable to recreate your problem so I want to see if what you have generated is a correct version of the schema at least.

**Reply by jccava** [reply]
`nwbOutputFile.nwb_version` yields `2.4.0`!  OK.  How do I fix it?

**Reply by jccava** [reply]
and 2.4.0 was the previous version.  I did `generateCore `after updating...

**Reply by lawrence-mbf** [reply]
Okay, if you wish to stick with Schema version `2.4.0` that is fine but it does look like you generated your files using an older version of MatNWB. Calling `generateCore('2.4.0');` should regenerate the right files with the correct validation.

**Reply by jccava** [reply]
Is there a reason not to go to 2.6.0?

**Reply by lawrence-mbf** [reply]
Oh, if you did call `generateCore` after updating then I will have to ask how you upgraded your MatNWB version?

**Reply by lawrence-mbf** [reply]
I cannot make a recommendation as that has to do with the NWB schema and not the MatNWB version (which is technically decoupled though it is aware of new schema versions). I would imagine that using the newest version would be good if you are starting out for writing.

**Reply by jccava** [reply]
Two different ways.  First I cloned from GitHub and used the Inno installer.  After I got these errors, I just downloaded the .zip and copied over the archive contents.

**Reply by lawrence-mbf** [reply]
Can you type `which NwbFile` into the command window? The Inno installer does add MatNWB to PATH.

**Reply by jccava** [reply]
I may have messed up things by nonchalantly moving between different versions.  Typing `which NwbFile` points to `blah\blah\Documents\MATLAB\James\MatNWB\NwbFile.m`.  In trying to roll back, I have just been moving code in and out of the `MatNWB` folder that the Inno installer created.  Sounds like I may have confused my install?  Should I just wipe it and reinstall 2.4.0?

**Reply by lawrence-mbf** [reply]
Yes, I would advise not using the Inno installer as well to ensure that everything installed correctly. It is also important to check that the MATLAB path is pointing to the correct version of MatNWB if you're using multiple versions (MatNWB will default to the directory on PATH to generate their class files).

**Reply by lawrence-mbf** [reply]
Oh, sorry, please wipe and install the newest release of MatNWB. It should support all schema versions older than what it states.

**Reply by jccava** [reply]
Thank you.  Should I generate the latest core, or 2.4.0 schema?

**Reply by lawrence-mbf** [reply]
Neither should have this issue anyway. For your needs, I would suggest using the most updated schema.

**Reply by jccava** [reply]
Thousands of thanks.  Errors gone.  --James
